salt/transport/ipc.py

```diff
@@ -615,147 +721,60 @@ class IPCMessageSubscriber(IPCClient):
-                if self._read_stream_future is None:
-                    self._read_stream_future = self.stream.read_bytes(4096, partial=True)
-
-                if timeout is None:
-                    wire_bytes = yield self._read_stream_future
-                else:
-                    future_with_timeout = FutureWithTimeout(
-                        self.io_loop, self._read_stream_future, timeout)
-                    wire_bytes = yield future_with_timeout
+    def __init__(self, socket_path, io_loop=None):
+        self.service = IPCMessageSubscriberService(socket_path, io_loop)
+        self.queue = tornado.queues.Queue()
 
-                self._read_stream_future = None
+    def connected(self):
+        return self.service.connected()
 
-                # Remove the timeout once we get some data or an exception
-                # occurs. We will assume that the rest of the data is already
-                # there or is coming soon if an exception doesn't occur.
-                timeout = None
+    def connect(self, callback=None, timeout=None):
+        return self.service.connect(callback=callback, timeout=timeout)
 
-                self.unpacker.feed(wire_bytes)
-                first = True
-                for framed_msg in self.unpacker:
-                    if first:
-                        ret = framed_msg['body']
-                        first = False
-                    else:
-                        self.saved_data.append(framed_msg['body'])
-                if not first:
-                    # We read at least one piece of data
-                    break
-        except TornadoTimeoutError:
-            # In the timeout case, just return None.
-            # Keep 'self._read_stream_future' alive.
-            ret = None
-        except tornado.iostream.StreamClosedError as exc:
-            log.trace('Subscriber disconnected from IPC %s', self.socket_path)
-            self._read_stream_future = None
-            exc_to_raise = exc
-        except Exception as exc:
-            log.error('Exception occurred in Subscriber while handling stream: %s', exc)
-            self._read_stream_future = None
-            exc_to_raise = exc
+    @tornado.gen.coroutine
+    def _feed(self, msgs):
+        for msg in msgs:
+            yield self.queue.put(msg)
 
-        if self._sync_ioloop_running:
-            # Stop the IO Loop so that self.io_loop.start() will return in
-            # read_sync().
-            self.io_loop.spawn_callback(self.io_loop.stop)
+    @tornado.gen.coroutine
+    def read_async(self, callback, timeout=None):
+        '''
+        Asynchronously read messages and invoke a callback when they are ready.
 
-        if exc_to_raise is not None:
-            raise exc_to_raise  # pylint: disable=E0702
-        self._sync_read_in_progress.release()
-        raise tornado.gen.Return(ret)
+        :param callback: A callback with the received data
+        '''
+        self.service.read(self)
```

Comment:
Isn't the service read also a coroutine?
Missing another yield?

Reply:
The same as in the previous comment.